### PR TITLE
[Feat] Add warning when user overrides log function in Callbacks.

### PR DESCRIPTION
**Pull request by tchaton**
## What does this PR do?

Fixes # (issue)

## Before submitting
- [x] Was this discussed/approved via a Github issue? (no need for typos and docs improvements)
- [x] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), Pull Request section?
- [x] Did you make sure your PR does only one thing, instead of bundling different changes together? Otherwise, we ask you to create a separate PR for every change.
- [ ] Did you make sure to update the documentation with your changes?
- [ ] Did you write any new necessary tests? 
- [ ] Did you verify new and existing tests pass locally with your changes?
- [ ] If you made a notable change (that affects users), did you update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/CHANGELOG.md)?

## PR review
Anyone in the community is free to review the PR once the tests have passed.
Before you start reviewing make sure you have read [Review guidelines](https://github.com/PyTorchLightning/pytorch-lightning/wiki/Review-guidelines). In short, see the following bullet-list:  

 - [ ] Is this pull request ready for review? (if not, please submit in draft mode)
 - [ ] Check that all items from **Before submitting** are resolved
 - [ ] Make sure the title is self-explanatory and the description concisely explains the PR
 - [ ] Add labels and milestones (and optionally projects) to the PR so it can be classified; _Bugfixes should be including in bug-fix release milestones (m.f.X) and features should be included in (m.X.b) releases._
 

## Did you have fun?
Make sure you had fun coding 🙃

### CODE REVIEW

File: pytorch_lightning/callbacks/base.py [hunk 1]
```diff
@@ -178,3 +178,19 @@ def on_before_zero_grad(self, trainer, pl_module, optimizer):
...
         """
         pass
+
+    def log(self, *_, **__):
+        """
+        This function is a placeholder. Don't override this function.
+        When using Lightning Trainer, the log function from
+        the model will be attached to this callback.
+        """
+        raise NotImplementedError
```
Comment (Borda) [inline pytorch_lightning/callbacks/base.py:188]:
```suggestion
        pass
```
Does this tell me as a user that I shall overwrite it instead of do not touch it?

shall we rather check (in the time of assignment log from model to callback) if the method log was overwritten and if so warn the user that his call is lost...

UPDATE: I see the overwritten check bellow

Comment (tchaton) [reply pytorch_lightning/callbacks/base.py:188]: Yes, I added the warning.

File: pytorch_lightning/utilities/model_helpers.py [hunk 2]
```diff
@@ -41,3 +46,10 @@ def is_overridden(method_name: str, model: Union[LightningModule, LightningDataM
     else:
         is_overridden = instance_attr.__code__ is not super_attr.__code__
     return is_overridden
+
+
+def at_least_one_overriden(method_names: List[str], current_object: Any, super_object: Any) -> bool:
+    is_overriden = False
+    for method_name in method_names:
+        is_overriden |= is_overridden_general(method_name, current_object, super_object)
```
Comment (SkafteNicki) [inline pytorch_lightning/utilities/model_helpers.py:54]:
```suggestion
    is_overriden = any(is_overridden_general(method_name, current_object, super_object) for method_name in method_names)
```

File: pytorch_lightning/callbacks/base.py [hunk 3]
```diff
@@ -178,3 +178,19 @@ def on_before_zero_grad(self, trainer, pl_module, optimizer):
...
         """
         pass
+
+    def log(self, *_, **__):
+        """
+        This function is a placeholder. Don't override this function.
+        When using Lightning Trainer, the log function from
+        the model will be attached to this callback.
+        """
+        pass
```
Comment (SkafteNicki) [inline pytorch_lightning/callbacks/base.py:188]: Why does one method `pass` and the other `raise NotImplementedError`?

File: pytorch_lightning/trainer/connectors/callback_connector.py [hunk 4]
```diff
@@ -106,5 +107,11 @@ def _trainer_has_checkpoint_callbacks(self):
 
     def attach_model_logging_functions(self, model):
         for callback in self.trainer.callbacks:
+            if at_least_one_overriden(["log", "log_dict"], callback, Callback):
```
Comment (SkafteNicki) [inline pytorch_lightning/trainer/connectors/callback_connector.py:110]: what would happen if the users has a attribute call `log` (or `log_dict`). We still override, but would the user also be warned in this case?

File: pytorch_lightning/callbacks/base.py (hunk 3 above)
Comment (tchaton) [reply pytorch_lightning/callbacks/base.py:188]: Done !

File: pytorch_lightning/trainer/connectors/callback_connector.py (hunk 4 above)
Comment (tchaton) [reply pytorch_lightning/trainer/connectors/callback_connector.py:110]: Good question. I don't except user having a property called `log` or `log_dict`, but if they do. It should return an error as we can't assign to property without a setter.

Comment (SkafteNicki) [reply pytorch_lightning/trainer/connectors/callback_connector.py:110]: Alright, fine by me :]

Comment (tchaton) [reply pytorch_lightning/trainer/connectors/callback_connector.py:110]: Do you think we should add a check on vars too ?

Comment (SkafteNicki) [reply pytorch_lightning/trainer/connectors/callback_connector.py:110]: No, I consider this a very edge case.

**Review by Borda — COMMENTED** [review]
pls add tests that with overwrite `log` warning is raised

File: pytorch_lightning/callbacks/base.py (hunk 3 above)
Comment (awaelchli) [inline pytorch_lightning/callbacks/base.py:188]: I still don't understand why this is just a placeholder. Why can't we make the logging call from here? Why is it necessary to delay and dynamically set the method? 
If this was just a regular method you wouldn't need to do all the strange warning messages and checks for overriding.

Comment (Borda) [reply pytorch_lightning/callbacks/base.py:188]: well, we probably could as each callback has a trainer, and each trainer has a ref to model... we can just raise an error if the user tries to use it before the model is linked...

Comment (awaelchli) [reply pytorch_lightning/callbacks/base.py:188]: Yes, something like that is what I mean exactly.

Comment (tchaton) [reply pytorch_lightning/callbacks/base.py:188]: Hey @awaelchli,

Users can currently do `pl_module.log(...)` in their callback functions, but it makes sense to be able to do `self.log` everywhere.

The log function is tied to the model. It will save the logged metrics in the `model.results = Result()` object, and then will save its reference to EpochResultStore.    

To enable what you want, we would need to move log function to trainer for both model and callbacks to call it internally, move results object to the trainer, add trainer reference to each callbacks, still make sure people don't override log and log_dict. 
It is possible by involve a slightly bigger refactor than attaching the log function to callbacks.

If you feel it isn't right, let's merge this PR as log functions are already attached and work on a refactor.

Best regards,
T.C

Comment (tchaton) [reply pytorch_lightning/callbacks/base.py:188]: We will do this in another PR. We will need to move logging to the Trainer, so it is cleaner.

Comment (Borda) [reply pytorch_lightning/callbacks/base.py:188]: I would rather do it in this one as this seems tobe the main reason for this PR :]
